File: cpp/test/TestAPI.cpp

#include "ParserSession.h"

#include "gtest/gtest.h"

#include <sstream>

class ParserSession;

using ParserSessionPtr = ParserSession *;


class APITest : public ::testing::Test {
protected:
    
    static ParserSessionPtr session;
    
    static void SetUpTestSuite() {
        
        session = new ParserSession();
    }
    
    static void TearDownTestSuite() {
        
        delete session;
    }
    
    void SetUp() override {
        
    }
    
    void TearDown() override {
        
    }
};

ParserSessionPtr APITest::session;


//
// this used to assert
//
TEST_F(APITest, Bug1) {
    
    auto strIn = std::string("abc[]");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());

    auto bufAndLen = BufferAndLength(str, strIn.size());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(bufAndLen, nullptr, opts);

    auto N = session->concreteParseLeaf(STRINGIFYMODE_NORMAL);

    session->releaseNodeContainer(N);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

//
// this used to hang
//
TEST_F(APITest, Hang1) {
    
    auto strIn = std::string("<<rr[R");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());

    auto bufAndLen = BufferAndLength(str, strIn.size());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(bufAndLen, nullptr, opts);

    auto N = session->parseExpressions();

    session->releaseNodeContainer(N);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

//
// this used to crash
//
TEST_F(APITest, Crash1) {
    
    auto strIn = std::string("0^^");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());

    auto bufAndLen = BufferAndLength(str, strIn.size());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(bufAndLen, nullptr, opts);

    auto N = session->parseExpressions();

    session->releaseNodeContainer(N);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

//
// this used to crash
//
TEST_F(APITest, Crash2) {
    
    auto strIn = std::string(".2^^0");
    
    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    auto bufAndLen = BufferAndLength(str, strIn.size());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(bufAndLen, nullptr, opts);
    
    auto N = session->parseExpressions();
    
    session->releaseNodeContainer(N);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();
    
    SUCCEED();
}

//
// this used to crash
//
TEST_F(APITest, Crash3) {
    
    auto strIn = std::string("12^^a.a");
    
    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    auto bufAndLen = BufferAndLength(str, strIn.size());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(bufAndLen, nullptr, opts);
    
    auto N = session->parseExpressions();
    
    session->releaseNodeContainer(N);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();
    
    SUCCEED();
}

//
// this used to crash
//
TEST_F(APITest, Crash4) {
    
    auto strIn = std::string("12..");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());

    auto bufAndLen = BufferAndLength(str, strIn.size());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(bufAndLen, nullptr, opts);

    auto N = session->parseExpressions();

    session->releaseNodeContainer(N);

    EXPECT_EQ(session->nonFatalIssues.size(), 1u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

//
// this used to crash
//
TEST_F(APITest, Crash5) {
    
    auto strIn = std::string("123\\\n.45");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());

    auto bufAndLen = BufferAndLength(str, strIn.size());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(bufAndLen, nullptr, opts);

    auto N = session->parseExpressions();

    session->releaseNodeContainer(N);

    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

//
// this used to crash
//
TEST_F(APITest, Crash6) {
    
    auto strIn = std::string("\\0560");
    
    auto str = reinterpret_cast<Buffer>(strIn.c_str());
    
    auto bufAndLen = BufferAndLength(str, strIn.size());
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_LINECOLUMN;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(bufAndLen, nullptr, opts);
    
    auto N = session->parseExpressions();
    
    session->releaseNodeContainer(N);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();
    
    SUCCEED();
}

//
// this used to crash
//
TEST_F(APITest, Crash7) {
    
    auto strIn = std::string("1+1");

    auto str = reinterpret_cast<Buffer>(strIn.c_str());

    auto bufAndLen = BufferAndLength(str, strIn.size());
    
    //
    // this was originally using SOURCECONVENTION_UNKNOWN, which was 0
    // but 0 is now SOURCECONVENTION_LINECOLUMN
    // so make up a bogus SourceConvention of 2
    //
    
    ParserSessionOptions opts;
    opts.srcConvention = static_cast<SourceConvention>(2);
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    auto res = session->init(bufAndLen, nullptr, opts);
    
    EXPECT_EQ(res, PARSERSESSIONINIT_ERROR);

    SUCCEED();
}

//
// this used to crash
//
// CODETOOLS-62
//
TEST_F(APITest, Crash8) {
    
    const unsigned char arr[] = {'(', '*', '\r', '\n', '*', ')'};

    auto bufAndLen = BufferAndLength(arr, 6);
    
    ParserSessionOptions opts;
    opts.srcConvention = SOURCECONVENTION_SOURCECHARACTERINDEX;
    opts.tabWidth = DEFAULT_TAB_WIDTH;
    opts.firstLineBehavior = FIRSTLINEBEHAVIOR_NOTSCRIPT;
    opts.encodingMode = ENCODINGMODE_NORMAL;
    opts.alreadyHasEOFSentinel = false;
    
    session->init(bufAndLen, nullptr, opts);

    auto N = session->parseExpressions();

    session->releaseNodeContainer(N);
    
    EXPECT_EQ(session->nonFatalIssues.size(), 0u);
    EXPECT_EQ(session->fatalIssues.size(), 0u);
    
    session->deinit();

    SUCCEED();
}

File: crate/src/tests/test_api.rs
use crate::{
    EncodingMode, FirstLineBehavior, ParserSession, SourceConvention, StringifyMode,
    DEFAULT_TAB_WIDTH,
};


//
// this used to assert
//
#[test]
fn APITest_Bug1() {
    let strIn = "abc[]";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.concreteParseLeaf(StringifyMode::Normal);

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

//
// this used to hang
//
#[test]
fn APITest_Hang1() {
    let strIn = "<<rr[R";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.parseExpressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

//
// this used to crash
//
#[test]
fn APITest_Crash1() {
    let strIn = "0^^";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.parseExpressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

//
// this used to crash
//
#[test]
fn APITest_Crash2() {
    let strIn = ".2^^0";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.parseExpressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

//
// this used to crash
//
#[test]
fn APITest_Crash3() {
    let strIn = "12^^a.a";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.parseExpressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

//
// this used to crash
//
#[test]
fn APITest_Crash4() {
    let strIn = "12..";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.parseExpressions();

    assert_eq!(session.nonFatalIssues().len(), 1);
    assert_eq!(session.fatalIssues().len(), 0);
}

//
// this used to crash
//
#[test]
fn APITest_Crash5() {
    let strIn = "123\\\n.45";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.parseExpressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

//
// this used to crash
//
#[test]
fn APITest_Crash6() {
    let strIn = "\\0560";

    let mut session = ParserSession::new(
        strIn.as_bytes(),
        SourceConvention::LineColumn,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.parseExpressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}

//
// this used to crash
//
// NOTE: This test was part of the C++ version, but is not possible in the
//       Rust version of CodeParser, which does not allow invalid values
//       for the SourceConvention.
#[test]
fn APITest_Crash7() {
    //let strIn = "1+1";

    //
    // this was originally using SOURCECONVENTION_UNKNOWN, which was 0
    // but 0 is now SourceConvention::LineColumn
    // so make up a bogus SourceConvention of 2
    //

    //let mut session = ParserSession::new(strIn.as_bytes(),static_cast<SourceConvention>(2), DEFAULT_TAB_WIDTH, FirstLineBehavior::NotScript, EncodingMode::Normal);
    //assert_eq!(res, PARSERSESSIONINIT_ERROR);
}

//
// this used to crash
//
// CODETOOLS-62
//
#[test]
fn APITest_Crash8() {
    const arr: &[u8] = &[b'(', b'*', b'\r', b'\n', b'*', b')'];

    let bufAndLen = arr;

    let mut session = ParserSession::new(
        bufAndLen,
        SourceConvention::CharacterIndex,
        DEFAULT_TAB_WIDTH,
        FirstLineBehavior::NotScript,
        EncodingMode::Normal,
    );

    let _ = session.parseExpressions();

    assert_eq!(session.nonFatalIssues().len(), 0);
    assert_eq!(session.fatalIssues().len(), 0);
}
